common/types/provider.go

```diff
@@ -302,237 +246,280 @@ var (
+		return Int(v), true
 	case int64:
-		return Int(value.(int64))
+		return Int(v), true
 	case uint:
-		return Uint(value.(uint))
+		return Uint(v), true
 	case uint32:
-		return Uint(value.(uint32))
+		return Uint(v), true
 	case uint64:
-		return Uint(value.(uint64))
+		return Uint(v), true
 	case float32:
-		return Double(value.(float32))
+		return Double(v), true
 	case float64:
-		return Double(value.(float64))
+		return Double(v), true
 	case string:
-		return String(value.(string))
+		return String(v), true
+	case time.Duration:
+		return Duration{Duration: v}, true
+	case time.Time:
+		return Timestamp{Time: v}, true
 	case *bool:
-		if ptr := value.(*bool); ptr != nil {
-			return Bool(*ptr)
+		if v != nil {
+			return Bool(*v), true
 		}
 	case *float32:
-		if ptr := value.(*float32); ptr != nil {
-			return Double(*ptr)
+		if v != nil {
+			return Double(*v), true
 		}
 	case *float64:
-		if ptr := value.(*float64); ptr != nil {
-			return Double(*ptr)
+		if v != nil {
+			return Double(*v), true
 		}
 	case *int:
-		if ptr := value.(*int); ptr != nil {
-			return Int(*ptr)
+		if v != nil {
+			return Int(*v), true
 		}
 	case *int32:
-		if ptr := value.(*int32); ptr != nil {
-			return Int(*ptr)
+		if v != nil {
+			return Int(*v), true
 		}
 	case *int64:
-		if ptr := value.(*int64); ptr != nil {
-			return Int(*ptr)
+		if v != nil {
+			return Int(*v), true
 		}
 	case *string:
-		if ptr := value.(*string); ptr != nil {
-			return String(*ptr)
+		if v != nil {
+			return String(*v), true
 		}
 	case *uint:
-		if ptr := value.(*uint); ptr != nil {
-			return Uint(*ptr)
+		if v != nil {
+			return Uint(*v), true
 		}
 	case *uint32:
-		if ptr := value.(*uint32); ptr != nil {
-			return Uint(*ptr)
+		if v != nil {
+			return Uint(*v), true
 		}
 	case *uint64:
-		if ptr := value.(*uint64); ptr != nil {
-			return Uint(*ptr)
+		if v != nil {
+			return Uint(*v), true
 		}
 	case []byte:
-		return Bytes(value.([]byte))
+		return Bytes(v), true
 	case []string:
-		return NewStringList(a, value.([]string))
+		return NewStringList(a, v), true
+	case []ref.Val:
+		return NewRefValList(a, v), true
 	case map[string]string:
-		return NewStringStringMap(a, value.(map[string]string))
-	case *dpb.Duration:
-		if ptr := value.(*dpb.Duration); ptr != nil {
-			return Duration{ptr}
+		return NewStringStringMap(a, v), true
+	case map[string]interface{}:
+		return NewStringInterfaceMap(a, v), true
+	case map[ref.Val]ref.Val:
+		return NewRefValMap(a, v), true
+	case *anypb.Any:
+		if v == nil {
+			return UnsupportedTypeConversionErr(v), true
 		}
-	case *structpb.ListValue:
-		if ptr := value.(*structpb.ListValue); ptr != nil {
-			return NewJSONList(a, ptr)
+		unpackedAny, err := v.UnmarshalNew()
+		if err != nil {
+			return NewErr("anypb.UnmarshalNew() failed for type %q: %v", v.GetTypeUrl(), err), true
 		}
-	case structpb.NullValue, *structpb.NullValue:
-		return NullValue
+		return a.NativeToValue(unpackedAny), true
```

Comment:
If an error occurs here, it will be returned to the caller which is the desired behavior. Fixing this behavior to more clearly spell out an error upon conversion might be desired, but I'm not sure it's needed and it would be a breaking change.